src/dunst.c

```diff
@@ -173,61 +172,39 @@ void wake_up(void)
         run(NULL);
 }
 
-static int get_sleep_time(void)
+static gint64 get_sleep_time(void)
 {
+        gint64 time = g_get_monotonic_time();
+        gint64 sleep = G_MAXINT64;
 
-        if (settings.show_age_threshold == 0) {
-                /* we need to update every second */
-                return 1;
-        }
-
-        bool have_ttl = false;
-        int min_ttl = 0;
-        int max_age = 0;
         for (GList *iter = g_queue_peek_head_link(displayed); iter;
                         iter = iter->next) {
                 notification *n = iter->data;
+                gint64 ttl = n->timeout - (time - n->start);
 
-                max_age = MAX(max_age, notification_get_age(n));
-                int ttl = notification_get_ttl(n);
-                if (ttl >= 0) {
-                        if (have_ttl) {
-                                min_ttl = MIN(min_ttl, ttl);
-                        } else {
-                                min_ttl = ttl;
-                                have_ttl = true;
-                        }
-                }
-        }
+                if (ttl >= 0)
+                        sleep = MIN(sleep, ttl);
```

Comment:
Since it came up in another comment: You should also handle separately the case of `n->timeout <= 0`, since that will also generate a negative `ttl` but for a different reason (notification doesn't expire).